### PR TITLE
Save disabled phases in `recipe.yaml`

### LINKED ISSUE #5059: `tmt run --last report -vvv` changes the recipe.yaml for that last run
Seems like it shouldn't. 

Though looking at the help, you could say that the behavior is expected.
```
$ tmt run --help | grep last
  -l, --last                      Execute the last run once again.
```
Maybe we can think about multiple recipe.yamls in the same run directory??

---

Reproducer:
```
$ tmt init --template base
$ tmt run
```
It should fail, but that's not the issue.

Copy the run recipe.yaml as a backup
```
$ cp /var/tmp/tmt/run-xxx/recipe.yaml /var/tmp/tmt/run-xxx/recipe.yaml.bak
```

Run the offending command:
```
$ tmt run --last report -vvv
```
Then diff the recipe.yamls. The new recipe.yaml only contains the report phase since that was just run.

### PR DESCRIPTION
**Pull request by therazix**
tmt now correctly saves disabled phases in the `recipe.yaml` if they contain valid data. For invalid data, tmt will show a warning and exclude the phases.

Also, tmt now saves the original `require` and `recommend` into `tests.yaml`, and the CLI context into `run.yaml`.

Fixes #5059 

Assisted-by: Cursor

Pull Request Checklist

* [x] implement the feature

### CODE REVIEW

**Review by LecrisUT — APPROVED** [review]

**Comment by psss** [comment]
Testing sync with Jira. Please, ignore.

File: tmt/recipe.py [hunk 1]
```diff
@@ -595,7 +598,7 @@ def save(self, run: 'Run') -> None:
             ),
             plans=[_RecipePlan.from_plan(plan) for plan in run.plans],
         )
-        self.write(run.run_workdir / 'recipe.yaml', tmt.utils.to_yaml(recipe.to_spec()))
+        self._save_recipe(run.run_workdir, recipe)
```
Comment (psss) [inline tmt/recipe.py:601]: I'm not completely sure this issue should/could be resolved by creating more and more copies of the `recipe.yaml` file. The real problem is that some of the steps are missing completely the data needed for the full re-execution. A simple scenario to verify:

    cd tests/execute/basic/data
    tmt run plan --name beakerlib
    tmt run --last report --how html --open

Now the user would like to execute the whole run according to the last modification, e.g. running all tests again and producing the `html` report, but many of the steps have empty phases in `recipe2.yaml` and the `html` report step is missing in `recipe.yaml`.

Comment (LecrisUT) [reply tmt/recipe.py:601]: If the first run is not intended to be used via recipe, that smells like a user error and probably shouldn't bother supporting. Too many loose ends like what if the user has a workflow like the traditional testing-farm reproducer with `since`, `login`, `until`, or worse.

This at least saves the user from accidentally loosing the recipe when they do the `run --last report`.

If they use it to inject reportportal and need to merge them, well that's a future issue, which we could probably resolve with a logic for merging the recipes.

Comment (therazix) [reply tmt/recipe.py]: > The real problem is that some of the steps are missing completely the data needed for the full re-execution.

That was intentional due to https://github.com/teemtee/tmt/pull/4661#discussion_r2986326496. But I think we can include them if they have valid data. Otherwise, we raise an exception for enabled steps, and for disabled steps, we just warn and exclude the phases. I made the changes in 241f680972b18f8568f1420d4003378d6b60f6d1. Does it make sense?

Comment (thrix) [reply tmt/recipe.py]: from hacking meeting:

* list use cases for the more complicate workflows
* outline the solution for the single recipe (rewriting), we should make sure we cover all corner cases if we choose the rewriting of the recipe
* if ^ is too complicated, fallback to multiple recipes

Comment (LecrisUT) [reply tmt/recipe.py]: Here are some of my scenarios of concern:

#### Manually injected phases should be present

```console
$ tmt run --until prepare
$ tmt run --last prepare --how shell ...
$ tmt run --last --since prepare
```
Here I would expect the recipe to include the dynamic phase that was added in `tmt run prepare --how shell` as well as all the other static ones that were run from the fmf files. There are other pitfalls like if the user used different context and environments, but I am just going to ignore that.

#### Running `tmt run --last report` should not invalidate the recipe

```console
$ tmt run ...
$ tmt run --last report ...
```

This is the scenario of #5059. We have a perfectly functioning run, but the user forgot to make a copy of the recipe, and they are annoyed that now they have to re-run everything from scratch. The recipe should show all the steps in the last run, even the previous `report` state that it had, because why not.

#### Disabled/skipped steps should be visible

```console
$ tmt run --until finish
```

Here I expect to see that the `finish` step was not executed.

#### Previous recipes from unrelated runs do not pollute the current run

```console
$ tmt run --id my_workspace --scratch
```

Here I expect the recipe state to be clean as if it was created anew.

Comment (tcornell-bus) [reply tmt/recipe.py]: After thinking about it more, my main concern is what Lecris is mentioning; I wouldn't want anything from my previous runs to be clobbered by future runs in the recipe.yaml file. As long as everything is kept (even if it's disabled), then it should work out.

Comment (therazix) [reply tmt/recipe.py]: I am not sure I understand what we expect the recipe to contain. I see the recipe as a static snapshot of a given tmt invocation, not as the full history of every re-execution of that workdir.

Recipe generation is based on the plans and steps loaded for that specific run. With `tmt run --last`, tmt reuses the same workdir. `run.yaml` is then rewritten with the steps enabled for the current invocation, and `step.yaml` files are updated for steps that run again. So rewriting a single recipe to match that same current state is consistent with how `--last` already behaves. Once a phase is overridden in `step.yaml`, the previous configuration is gone from the workdir, so I think the recipe should not try to merge it. To include phases from all re-executions, we would need to load the previous recipe and merge it, which I think would bring a lot of complications, for example, what to do when there are multiple phases with the same name? Or what to do with run level fields in the recipe?

Comment (therazix) [reply tmt/recipe.py]: > Here are some of my scenarios of concern:
> 
> #### Manually injected phases should be present
> ```
> $ tmt run --until prepare
> $ tmt run --last prepare --how shell ...
> $ tmt run --last --since prepare
> ```
> 
> Here I would expect the recipe to include the dynamic phase that was added in `tmt run prepare --how shell` as well as all the other static ones that were run from the fmf files. There are other pitfalls like if the user used different context and environments, but I am just going to ignore that.


The way it works now, the final recipe would have `prepare` through `cleanup` enabled and `discover` and `provision` disabled. It would still carry the original fmf prepare phases, not the CLI one, because after `tmt run --until prepare` the prepare step is already `done` and `tmt run --last prepare --how shell ...` does not change it without `--force`/`--again`. With `--force`/`--again` on that command, the final recipe would contain the new prepare phase, replacing the original one.


> #### Running `tmt run --last report` should not invalidate the recipe
> ```
> $ tmt run ...
> $ tmt run --last report ...
> ```
> 
> This is the scenario of #5059. We have a perfectly functioning run, but the user forgot to make a copy of the recipe, and they are annoyed that now they have to re-run everything from scratch. The recipe should show all the steps in the last run, even the previous `report` state that it had, because why not.

Currently (after the fix in 241f680972b18f8568f1420d4003378d6b60f6d1), it would contain the steps from the last run, but not the previous `report` phase.

> #### Disabled/skipped steps should be visible
> ```
> $ tmt run --until finish
> ```
> 
> Here I expect to see that the `finish` step was not executed.

The finish step **would** be executed. Report and cleanup would not be executed, and in the recipe, they would still have data, but these two last steps would be disabled.

> #### Previous recipes from unrelated runs do not pollute the current run
> ```
> $ tmt run --id my_workspace --scratch
> ```
> 
> Here I expect the recipe state to be clean as if it was created anew.

Yes, that should already work, because `--scratch` removes the workdir.

Comment (LecrisUT) [reply tmt/recipe.py]: > I see the recipe as a static snapshot of a given tmt invocation, not as the full history of every re-execution of that workdir.

To me it can be as either. If we go with the multiple recipe save, then having it be the former is perfectly fine. But if we keep only one recipe, then, it can be problematic that again we randomly loose information, the same as what the original issue mentions.

> To include phases from all re-executions, we would need to load the previous recipe and merge it, which I think would bring a lot of complications, for example, what to do when there are multiple phases with the same name? Or what to do with run level fields in the recipe?

Indeed and that's why I preferred the previous approach where we are not opinionated and the user figures out what to do. We just provide them with all the information they may need.

Comment (psss) [reply tmt/recipe.py]: Let's step back just a little bit to remind the main use case:

> As a user, I want tmt to generate a recipe file that can be reused to reproduce a run with the exact same configuration, without requiring additional input.

Recipe is something very close to the Beaker recipe xml, except for the fact that in Beaker there is always the whole job executed at once, no `--until` and no `--last`. So I see several cases we want to cover:

## Just one shot

One `tmt run`, from the beginning to the end, no adjustment in the middle or after the run has completed. All steps are enabled. Can be used in two ways:

* original recipe (just resubmit, exactly the same execution)
* modified recipe (e.g. filter errored or failed tests only)

Very similar to the Beaker usage. These two should cover the main use cases just well, including the CTC / compose qualification testing. @brakogh, please correct me if I'm wrong.

## Run with adjustments

The run is not executed in one shot, some modifications to the flow are done either during or after it is finished:

* after (using --last to generate additional report or add finish actions like submit logs)
* middle (run selected steps using --until, --before, --after, --since)

Some of the steps are enabled, some enabled. This is reflected in the recipe. But what are the exact expectations for rerunning such a recipe? Running again just a subset of the enabled steps? That does not seem to be right or very useful I'd say.

For example, the `tmt run --last report` case: When would a user want to schedule a new run which would only do a `report` step using the `recipe.yaml`? Does not make much sense to me. Or with `tmt run --since execute` rescheduling freshly the second half of the recipe? With no `provision` step this does not make any sense.

## Simplified approach

If we don't have clear use for such scenarios I would not invest too much effort in implementing something very complex. What about just running all steps by default when using `--recipe`, regardles the saved `enabled` status? If user wants only selected steps, they can explicitly specify them on the command line:

    tmt run --recipe recipe.yaml discover

This could simplify this part of the implementation. Regarding the question of what should happen with "updated" step phases: I would say it should behave consistently with how overriding works from command line: Any step phase touched during/after the run would be stored modified, the rest should stay untouched.

The following would replace fmf `prepare` step config with the command line one, original `prepare` step not executed at all thanks to `--before`:

    tmt run --before prepare
    tmt run --last prepare --how shell ...
    tmt run --last --after prepare

If `--insert` is used then the phase would be inserted in addition to the phases defined in the `fmf` file, just as it is done in the one-shot `tmt run` case:

    tmt run --before prepare
    tmt run --last prepare --insert --how shell ...
    tmt run --last --after prepare

So the "interrupted" scenario would be consistent with the one-shot scenario. No special handling. Just modify provided steps in the very same way. For the `tmt run --last` case, it would just update the step as from the command line, all other step phases would be untouched.

For the `tmt run --until finish` scenario, `finish` step would be enabled (as `--until` is inclusive) and `cleanup` step disabled.  But, when rescheduling based on `recipe.yaml`, these states would not be respected.

What do you think?

Comment (LecrisUT) [reply tmt/recipe.py]: > The following would replace fmf `prepare` step config with the command line one, original `prepare` step not executed at all thanks to `--before`:
> 
> ```
> tmt run --before prepare
> tmt run --last prepare --how shell ...
> tmt run --last --after prepare
> ```

This is not the scenario that I posed before. The one in question is if it used `--since prepare`, not `--after prepare`. Sure the scenario with `--after` is easy to give an answer to figure out a "correct answer", but the specific edge cases is where we actually figure out how to define and implement it. E.g. the "replace all steps with the ones that were executed" works fine for the `--after` case, but for the `--since`, the `--how shell` stuff is lost. And I could equally propose that recipes should represent all phases that were executed in the run workdir, which would not introduce that replacement in the `--since` case.

We also have even more funky breaking scenarios as @therazix or I pointed with

> Or what to do with run level fields in the recipe?

> There are other pitfalls like if the user used different context and environments, but I am just going to ignore that.

---

Ultimately why do we care about such complications? Why shouldn't it be the user's responsibility to define how to consume. The original issue #5059 has a very specific scenario, save the user from shooting themselves in the foot, for which we only need to know the previous state(s) of the runs. If the users really want only one recipe file or they are confused why there are so many recipes, well, let's let them voice their thoughts and have them contribute to how this merger should happen, because as we can see, it is not a straightforward answer.

Comment (therazix) [reply tmt/recipe.py]: > If we don't have clear use for such scenarios I would not invest too much effort in implementing something very complex. What about just running all steps by default when using --recipe, regardles the saved enabled status? If user wants only selected steps, they can explicitly specify them on the command line:

So if I understand it correctly, we would ignore the enabled state of the steps and just execute the whole recipe by default? In that case, we could remove the enabled key entirely, since it would cause confusion and I don't see the point of having it.

> Some of the steps are enabled, some enabled. This is reflected in the recipe. But what are the exact expectations for rerunning such a recipe? Running again just a subset of the enabled steps? That does not seem to be right or very useful I'd say.

But it correctly reflects the last run. Users can modify the recipe to enable the steps they want. Are there any use cases where users would purposely skip some steps and would want the recipe to skip those steps as well? Maybe something like `tmt run --before cleanup` with no additional `tmt run --last ...` commands? It could be unwanted to execute the `cleanup` step by default using that recipe. But it depends on whether there are such use cases. If not, I have no problem with enabling all steps in the recipe by default.

> Regarding the question of what should happen with "updated" step phases: I would say it should behave consistently with how overriding works from command line: Any step phase touched during/after the run would be stored modified, the rest should stay untouched.

I agree, and this currently works that way, including `--insert`. tmt will load the previous run, updates existing phases based on the CLI (or inserts them with `--insert`), and saves them in the new recipe.

It does not save the full history in a single recipe. If we have something like this:
```
# Let's assume the prepare step from fmf contains only one phase
tmt run
tmt run --last --again prepare -h shell --script "echo 1"
tmt run --last --again prepare -h shell --script "echo 2"
tmt run --last --again prepare -h shell --script "echo 3"
```

The final recipe will have the data from fmf (the first command), and the `prepare` step will contain only one phase, with the script from the last execution (`echo 3`).

Comment (psss) [reply tmt/recipe.py]: > This is not the scenario that I posed before. The one in question is if it used `--since prepare`, not `--after prepare`. Sure the scenario with `--after` is easy to give an answer to figure out a "correct answer", but the specific edge cases is where we actually figure out how to define and implement it.

Yeah, I've simplified that use case, just as the first step, to agree / to get a rough idea if that approach would be acceptable over all. Ok, let's cover the usage with `--until` or `--since` as well:

* If the particular `prepare` step or steps defined in `fmf` metadata were already executed, e.g. running with `--until` and user comes and does additional `tmt run --last prepare ...` it can mean two things: Either there was something wrong (I want to override the wrong prepare `shell` phase with a correct one) or there are extra steps to be added (I want to add one more action which I forgot to include). As we are not able to discern the two I would suggest to rely on the current way how cli overrides/extends, that is override with `tmt run --last prepare` and extend with `tmt run --last prepare --insert`. Same would apply for the `--since` use case.

> > Or what to do with run level fields in the recipe?

Which specific fields are meant by this?

> > If we don't have clear use for such scenarios I would not invest too much effort in implementing something very complex. What about just running all steps by default when using --recipe, regardles the saved enabled status? If user wants only selected steps, they can explicitly specify them on the command line:
> 
> So if I understand it correctly, we would ignore the enabled state of the steps and just execute the whole recipe by default?

Yes, that's what I proposed/brainstormed.

> In that case, we could remove the enabled key entirely, since it would cause confusion and I don't see the point of having it.

Hm, good point. Yeah, seems like there would be no use case for the `enabled` field, other than informational only.

> But it correctly reflects the last run. Users can modify the recipe to enable the steps they want. Are there any use cases where users would purposely skip some steps and would want the recipe to skip those steps as well? Maybe something like `tmt run --before cleanup` with no additional `tmt run --last ...` commands?

That one actually sounds like a valid use case. Thinking of some more... maybe intentionally skipping some step? Like `--skip prepare` to just provision and run the set of tests on an unchanged guest. Hmmm... so there are some combinations which are useful and some which are clearly not?

> I agree, and this currently works that way, including `--insert`. tmt will load the previous run, updates existing phases based on the CLI (or inserts them with `--insert`), and saves them in the new recipe.

Good!

> It does not save the full history in a single recipe. If we have something like this:
> 
> ```
> # Let's assume the prepare step from fmf contains only one phase
> tmt run
> tmt run --last --again prepare -h shell --script "echo 1"
> tmt run --last --again prepare -h shell --script "echo 2"
> tmt run --last --again prepare -h shell --script "echo 3"
> ```
> 
> The final recipe will have the data from fmf (the first command), and the `prepare` step will contain only one phase, with the script from the last execution (`echo 3`).

That sounds like the expected behavior to me. Well... actually, not quite so: After `tmt run` there is no guest running as `cleanup` has stopped and removed any instance so `tmt run --last prepare --again` here does not make any sense.

Comment (therazix) [reply tmt/recipe.py]: > > > Or what to do with run level fields in the recipe?
> 
> Which specific fields are meant by this?

For example run environment or context.

> > But it correctly reflects the last run. Users can modify the recipe to enable the steps they want. Are there any use cases where users would purposely skip some steps and would want the recipe to skip those steps as well? Maybe something like `tmt run --before cleanup` with no additional `tmt run --last ...` commands?
> 
> That one actually sounds like a valid use case. Thinking of some more... maybe intentionally skipping some step? Like `--skip prepare` to just provision and run the set of tests on an unchanged guest. Hmmm... so there are some combinations which are useful and some which are clearly not?

Yes, someone might prefer to have all steps enabled by default when loading the recipe, while someone else might prefer to have only the steps enabled that were enabled in the last run. So the question is which of these approaches we should take.

> > It does not save the full history in a single recipe. If we have something like this:
> > ```
> > # Let's assume the prepare step from fmf contains only one phase
> > tmt run
> > tmt run --last --again prepare -h shell --script "echo 1"
> > tmt run --last --again prepare -h shell --script "echo 2"
> > tmt run --last --again prepare -h shell --script "echo 3"
> > ```
> > The final recipe will have the data from fmf (the first command), and the `prepare` step will contain only one phase, with the script from the last execution (`echo 3`).
> 
> That sounds like the expected behavior to me. Well... actually, not quite so: After `tmt run` there is no guest running as `cleanup` has stopped and removed any instance so `tmt run --last prepare --again` here does not make any sense.

Yes, running prepare here does not make much sense, maybe only with local provision. It was just an example to determine how the phases should be saved in the recipe. If you say this is expected behavior, I agree with that approach.

Comment (psss) [reply tmt/recipe.py]: > > I agree, and this currently works that way, including `--insert`. tmt will load the previous run, updates existing phases based on the CLI (or inserts them with `--insert`), and saves them in the new recipe.
> 
> Good!

I've just verified this is working as expected. So, if the main remaining discussion point is about creating the backup files, what about putting that functionality under a dedicated option? I suggest to keep a single recipe by default and enable creating backups using something like `--backup-recipe`. @therazix, what do you think?

Comment (LecrisUT) [reply tmt/recipe.py]: `--backup-recipe` is meaningless. Either it is too late to save the user, or the user would simply copy the file

Comment (psss) [reply tmt/recipe.py]: > Yes, someone might prefer to have all steps enabled by default when loading the recipe, while someone else might prefer to have only the steps enabled that were enabled in the last run. So the question is which of these approaches we should take.

Ack, so this is also a question about what should be the default:

* respect `recipe.yaml` and allow user to override, e.g. using `--all` or listing the desired steps
* run all steps by default and implement a new option to respect `enabled` in the `recipe.yaml`

Actually, adding `--all` is not that hard, in some scenarios it might be surprising that only subset of steps is run, but in most cases it should be ok. I'm leaning towards the first option above (meaning no additional code change is needed).

Comment (therazix) [reply tmt/recipe.py]: From the hacking session: Let's override the existing recipe on re-run, and respect the enabled state from the recipe during loading. It should be possible to override this using `--all` option or by specifying the steps on command line.

Comment (therazix) [reply tmt/recipe.py]: > From the hacking session: Let's override the existing recipe on re-run, and respect the enabled state from the recipe during loading. It should be possible to override this using `--all` option or by specifying the steps on command line.

I updated the PR. f943652fa06e1470629d497aee3b6c8475d02b67

**Review by LecrisUT — CHANGES_REQUESTED** [review]
There are edge-cases that this does not cover, but as long as the simple scenario of running `tmt run --last report` is covered, I am all fine with it. Can you add a test for it? At the very least there is something off when I try it:

<details>

```console
$ tmt --feeling-safe run -a prepare --insert --how=shell --script="echo hi" plan -n /plans/local
$ cp /var/tmp/tmt/run-XXX/recipe{,_initial}.yaml
$ tmt run --last report --how=html
$ cp /var/tmp/tmt/run-XXX/recipe{,_report}.yaml
$ diff /var/tmp/tmt/run-XXX/recipe_{initial,report}.yaml
3c3
<     enabled: true
---
>     enabled: false
9c9
<     enabled: true
---
>     enabled: false
54,56c54
<       - name: /very/deep/file
<         type: library
<         url: https://github.com/beakerlib/test
---
>       - coreutils
104c102
<     enabled: true
---
>     enabled: false
117c115
<     enabled: true
---
>     enabled: false
129c127
<     enabled: true
---
>     enabled: false
142c140
<     enabled: true
---
>     enabled: false
159,160c157,159
<     - display-guest: auto
<       how: display
---
>     - absolute-paths: false
>       display-guest: auto
>       how: html
161a161
>       open: false
165,167c165
<   context:
<     ctx:
<     - something
---
>   context: {}
```

- the requires in the `discover/fmf` are misaligned, looks like a list mutation issue somewhere
- context is not saved across recipes

</details>

**Comment by therazix** [comment]
> * the requires in the `discover/fmf` are misaligned, looks like a list mutation issue somewhere

The problem is that the recipe uses `_original_require/_original_recommend`, but `tests.yaml` only stores the evaluated package lists. With `--last`, tests are loaded from `tests.yaml`, and we no longer have the original values.

Would it make sense to also save the original require and recommend in `tests.yaml`?

> * context is not saved across recipes

Seems like similar issue. `context` is not saved in the `run.yaml`.

**Comment by therazix** [comment]
> Would it make sense to also save the original require and recommend in `tests.yaml`?

Added in efc1e192ff365f9fc99af15d9df075d678bfb377.

File: tmt/recipe.py (hunk 1 above)
Comment (psss) [reply tmt/recipe.py]: Thanks. Looks good and works as expected.

**Review by psss — APPROVED** [review]

**Review by LecrisUT — APPROVED** [review]
